### DISCUSSION TITLE
[Feature] Centralized Metadata Storage with Customizable Hashing and Enhanced Database Backup

### DISCUSSION THREAD

**Discussion by anthonynogales** [immich-app/immich#13896] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

Currently, Immich’s metadata handling seems to be limited to the creation of XMP sidecar files which is a problem in read-only libraries. While this approach provides some flexibility, a more centralized and robust metadata system applied across all library types could greatly improve usability, reliability, and security for users with different workflows and needs.

### Feature Updates

1. **Centralized Metadata Storage for All Libraries**  
   Store metadata (such as descriptions, tags, and ratings) in the database by default for all libraries, regardless of whether they are read-only or writable. This centralized repository would:
   - Enable efficient metadata organization without modifying original files
   - Allow users to optionally export metadata to XMP sidecar files, either for individual images or an entire library, based on preference
   - Provide a consistent and reliable approach to metadata handling across all types of libraries

2. **Customizable Hash Generation and Management**  
   Immich could add support for generating cryptographic hashes (MD5, SHA-256, SHA-512, etc.) for photos and videos. These hashes would be stored in the database alongside metadata, allowing users to:
   - Select their preferred hash type at setup or change it later, updating to a different hash type if desired
   - Verify new hash calculations against existing hashes (when switching types) to ensure accuracy
   - Optionally store multiple hash types per file, enhancing data integrity and giving users peace of mind regarding data consistency

3. **Compatibility with Read-Only Libraries**  
   Extending hash generation and metadata storage functionality to read-only libraries would provide users confidence that:
   - Immich will not modify their original files, thus preserving data integrity
   - They can use Immich strictly as a gallery and metadata management system without needing write permissions on files
   - Metadata can be exported if they later decide to make a library writable and switch to XMP sidecar storage

4. **In-App Database Backup and Restore Options**  
   To make database management simpler, Immich could benefit from an in-app option to export and import the database. This feature would:
   - Make backing up and restoring metadata, hashes, and other configurations easy and accessible without relying on CLI commands
   - Reduce the risk of error for less technical users while encouraging long-term database management within Immich
   - Help Immich serve as a comprehensive solution, especially for users transitioning from applications like Adobe Lightroom

With these enhancements, Immich could become a reliable, long-term solution for users who prioritize metadata control, data integrity, and security across all library types. By storing metadata and hash information in a centralized database and offering in-app backup, Immich would become a feature-rich platform for gallery management that prioritizes data integrity and security-conscious users.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by bo0tzz** [comment]
The template clearly requests that you search for duplicates (some of these have already been requested) and that you stick to one request per thread.

**Reply by anthonynogales** [reply]
Well, I hope you implement these features to make this a knockout app. Otherwise, they’re its Achilles' heel.
